spec.html

```diff
@@ -20700,22 +20705,17 @@ <h1>Runtime Semantics: ClassDefinitionEvaluation</h1>
         1. Let _proto_ be ! OrdinaryObjectCreate(_protoParent_).
         1. If |ClassBody_opt| is not present, let _constructor_ be ~empty~.
         1. Else, let _constructor_ be ConstructorMethod of |ClassBody|.
-        1. If _constructor_ is ~empty~, then
-          1. If |ClassHeritage_opt| is present, then
-            1. Let _constructorText_ be the source text
-              <pre><code class="javascript">constructor(...args) { super(...args); }</code></pre>
-          1. Else,
-            1. Let _constructorText_ be the source text
-              <pre><code class="javascript">constructor() {}</code></pre>
-          1. Set _constructor_ to ParseText(_constructorText_, |MethodDefinition[~Yield, ~Await]|).
-          1. Assert: _constructor_ is a Parse Node.
         1. Set the running execution context's LexicalEnvironment to _classScope_.
-        1. Let _constructorInfo_ be ! DefineMethod of _constructor_ with arguments _proto_ and _constructorParent_.
-        1. Let _F_ be _constructorInfo_.[[Closure]].
-        1. Perform SetFunctionName(_F_, _className_).
-        1. Perform MakeConstructor(_F_, *false*, _proto_).
+        1. If _constructor_ is ~empty~, then
+          1. Let _steps_ be the algorithm steps defined in <emu-xref href="#sec-default-constructor-functions"></emu-xref>.
+          1. Let _F_ be ! CreateBuiltinFunction(_steps_, 0, _className_, &laquo; [[ConstructorKind]], [[SourceText]] &raquo;, ~empty~, _constructorParent_).
```

Comment:
Why create `_F_` as a *built-in* function?

Reply:
I don't understand this question.

Reply:
Why create it as a built-in, rather than creating it as an ordinary (ECMAScript) function? It's an ordinary function in the status quo.

Reply:
To make the default constructor not depend on the `@@iterator` property of `%Array.prototype%` being unmodified.

Reply:
Is the difference observable? A class constructor’s toString is of the class, not the constructor function.

Reply:
@ljharb Only if you mutate `%Array.prototype%`’s `@@iterator` property.

---

Also, this PR makes it so that if a built‑in function has an `[[ECMAScriptCode]]` internal slot that contains a parse node, then `%Function.prototype.toString%` returns the text covered by the parse node.

Reply:
> if you mutate %Array.prototype%’s @@iterator property.

That's the point of this PR. Check the linked issue.

Reply:
@ExE-Boss the function being built-in, or ecmascript code, isn't related to that difference (which is indeed the point of the PR).

Reply:
@ljharb That’s why this PR makes said change to `%Function.prototype.toString%, so that:
```js
(class Foo extends Bar {}).toString();
```

keeps returning:
```js
"class Foo extends Bar {}"
```

Reply:
Okay, let me see if I've got this straight: there's no way to write an ECMAScript function that has the necessary behavior of a default constructor without it also having the undesirable behavior of accessing `Array.prototype[@@iterator]`. So an ECMAScript/ordinary function is out.

So that still leaves the question: if what you need is an exotic function, why use CreateBuiltinFunction? After all, implementations are allowed to implement built-in functions as ordinary or exotic, and we want to avoid ordinary. Presumably, by specifying semantics that can't be embodied in an ordinary function, you force the implementation to use an exotic function, but that seems somewhat indirect.
- So is it just convenience? I.e., using CreateBuiltinFunction (and specifying un-ordinary semantics) is just easier than specifying an exotic function "from scratch".
- Or is there something about the behavior of an exotic built-in that you're not guaranteed to get from an exotic _non_-built-in?

Reply:
I don't understand what you're saying. This is a builtin function. Builtin functions in the spec are created via CreateBuiltinFunction. Is there an alternative I'm not aware of here?

Reply:
@jmdyck Convenience. More specifically, the goal is to specify a function which behaves in a particular way, not to specify an exotic function. There's no good way to specify the behavior we want with ECMAScript code, and besides that's not how we usually specify functions, so we need to specify a function some other way. And CreateBuiltinFunction is the way we usually do that. That the specified behavior happens to force an implementation to create an exotic function is entirely incidental.

Reply:
Okay, I think I understand now.